firebase-firestore/src/test/java/com/google/firebase/firestore/local/BundleCacheTestCase.java

```diff
@@ -0,0 +1,130 @@
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNull;
+
+import com.google.firebase.Timestamp;
+import com.google.firebase.firestore.core.Query;
+import com.google.firebase.firestore.model.ResourcePath;
+import com.google.firebase.firestore.model.SnapshotVersion;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TestName;
+
+/**
+ * These are tests for any implementation of the BundleCache interface.
+ *
+ * <p>To test a specific implementation of BundleCache:
+ *
+ * <ol>
+ *   <li>Subclass BundleCacheTestCase.
+ *   <li>Override {@link #getPersistence}, creating a new implementation of Persistence.
+ * </ol>
+ */
+public abstract class BundleCacheTestCase {
+  @Rule public TestName name = new TestName();
+
+  private Persistence persistence;
+  private BundleCache bundleCache;
+
+  @Before
+  public void setUp() {
+    persistence = getPersistence();
+    bundleCache = persistence.getBundleCache();
+  }
+
+  @After
+  public void tearDown() {
+    persistence.shutdown();
+  }
+
+  abstract Persistence getPersistence();
+
+  @Test
+  public void testReturnsNullWhenBundleIdNotFound() {
+    assertNull(bundleCache.getBundleMetadata("bundle-1"));
+  }
+
+  @Test
+  public void testReturnsSavedBundle() {
+    BundleMetadata expectedBundleMetadata =
+        new BundleMetadata("bundle-1", 1, new SnapshotVersion(Timestamp.now()));
+    bundleCache.saveBundleMetadata(expectedBundleMetadata);
+    BundleMetadata actualBundleMetadata = bundleCache.getBundleMetadata("bundle-1");
+
+    assertEquals(expectedBundleMetadata, actualBundleMetadata);
+
+    // Overwrite
+    expectedBundleMetadata =
+        new BundleMetadata("bundle-1", 2, new SnapshotVersion(Timestamp.now()));
+    bundleCache.saveBundleMetadata(expectedBundleMetadata);
+    actualBundleMetadata = bundleCache.getBundleMetadata("bundle-1");
+
+    assertEquals(expectedBundleMetadata, actualBundleMetadata);
+  }
+
+  @Test
+  public void testReturnsNullWhenQueryNameIsNotFound() {
+    assertNull(bundleCache.getNamedQuery("query-1"));
+  }
+
+  @Test
+  public void testReturnsSavedCollectionQueries() {
+    NamedQuery expectedQuery =
+        new NamedQuery("query-1", query("collection"), new SnapshotVersion(Timestamp.now()));
+    bundleCache.saveNamedQuery(expectedQuery);
+
+    NamedQuery actualQuery = bundleCache.getNamedQuery("query-1");
+    assertEquals(expectedQuery, actualQuery);
+  }
+
+  @Test
+  public void testReturnsSavedLimitToFirstQueries() {
+    NamedQuery expectedQuery =
+        new NamedQuery(
+            "query-1", query("collection").limitToFirst(1), new SnapshotVersion(Timestamp.now()));
+    bundleCache.saveNamedQuery(expectedQuery);
+
+    NamedQuery actualQuery = bundleCache.getNamedQuery("query-1");
+    assertEquals(expectedQuery, actualQuery);
+  }
+
+  @Test
+  public void testReturnsSavedLimitToLastQueries() {
```

Comment:
Note the point of having a limit-to-last test is queries saved in bundles are encoded as target with an extra boolean saying it is limit to last. 

So I think constructing a limit to last query here defeat the purpose, checking the JS implementation, it is constructing a limit to first query with the extra boolean (by creating the proto class manually), to ensure limit to last queries in bundles are translated correctly.

Also check `local_serializer.ts/fromBundleQuery` to see the actual translation for limit queries.

Reply:
Since I updated the data layer to use only NamedQuery, I don't have a way to inject Protos in this test. I do have a test thought that likely serves the same purpose (`testEncodesLimitToLastQuery` in the file below). This test actually uncovered an issue I had during the development of this PR, since I originally called toTarget() before serialization which flipped the order by constraints.